### ISSUE TITLE
Question: Cross-encoder training with MultipleNegativesRankingLoss (MNR Loss)

### ISSUE THREAD

**Issue by sciencecw** [UKPLab/sentence-transformers#2805]
Is it possible to train a cross encoder with the same loss used to train a bi-encoder?

**Comment by tomaarsen** [comment]
Hello!

I'm afraid not. Cross-encoders and bi-encoders are fundamentally different. For bi-encoders you can get embeddings for your "anchors" and "positives", and then train to "push apart" the i-th anchor with the j-th positive (i.e., an "in-batch negative pair") in the embedding space. These in-batch negative pairs can be created really cheaply for bi-encoders once you've already embedded the anchors and positives, but for a cross-encoder you would need one forward call for each negative pair. 

In short: no, you can't use the same code because bi-encoders and cross-encoders work differently (e.g. inference on 1 text at a time VS inference on 1 pair at a time), *but* you can use the same approach of in-batch negatives, I believe.

- Tom Aarsen

**Comment by sciencecw** [comment]
I understand now.

I notice that the negative mining utility just got merged (#2768) and I am going to give it a try

**Comment by tomaarsen** [comment]
Feel free to update me on whether it works well for you!
To give you an idea, here were some mined datasets that I uploaded:
- https://huggingface.co/datasets/tomaarsen/gooaq-hard-negatives
- https://huggingface.co/datasets/tomaarsen/natural-questions-hard-negatives

---

- Tom Aarsen
